Gradle Node Plugin fails in some use case.

<!--
- Please follow the issue template below for bug reports.
- If you have a support request rather than a bug, please use [Stack Overflow](http://stackoverflow.com/questions/tagged/jhipster) with the JHipster tag.
-
- If this is a potential security issue, please read our security policy and contact us privately first: https://github.com/jhipster/generator-jhipster/security/policy
-->

##### **Overview of the issue**

```
[backend] Error: Cannot find module '../lib/cli.js'
[backend] Require stack:
[backend] - /Users/runner/work/generator-jhipster/generator-jhipster/app/build/node/bin/npm
[backend]     at Function._resolveFilename (node:internal/modules/cjs/loader:1249:15)
[backend]     at Function._load (node:internal/modules/cjs/loader:1075:27)
[backend]     at TracingChannel.traceSync (node:diagnostics_channel:322:14)
[backend]     at wrapModuleLoad (node:internal/modules/cjs/loader:219:24)
[backend]     at Module.require (node:internal/modules/cjs/loader:1340:12)
[backend]     at require (node:internal/modules/helpers:138:16)
```

Reproduction:
- Convert deserver sample to gradle: https://github.com/jhipster/generator-jhipster/pull/28163/commits/3a046edd0800ed2b6b83072838bbd54c61d63a5b

<!-- Explain the bug, if an error is being thrown a stack trace helps -->

##### **Motivation for or Use Case**

<!-- Explain why this is a bug for you -->

##### **Reproduce the error**

<!-- For bug reports, an unambiguous set of steps to reproduce the error -->

##### **Related issues**

<!-- Has a similar issue been reported before? Please search both closed & open issues -->

##### **Suggest a Fix**

<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**

<!--
Which version of JHipster are you using, is it a regression?
-->

##### **Browsers and Operating System**

<!-- What OS are you on? is this a problem with all browsers or only IE8? -->

- [ ] Tickets opened without reproduction steps or that doesn't follows the template recommendation will be **closed**.
- [ ] This issue is prompt-related or an error that prevents JHipster from generating an application.
- [ ] I don't have a JDL otherwise I should open an [JDL Issue](https://github.com/jhipster/generator-jhipster/issues/new?template=BUG_REPORT_JDL.md)
- [ ] The application is not successfully generated otherwise, I should open an [Issue with jhipster info](https://github.com/jhipster/generator-jhipster/issues/new?template=BUG_REPORT.md)
- [ ] Checking this box is mandatory (this is just to show you read everything)

<!-- Love JHipster? Please consider supporting our collective:
👉  https://opencollective.com/generator-jhipster/donate -->

--- a/generators/java/generators/node/templates/buildSrc/src/main/groovy/jhipster.node-gradle-conventions.gradle.ejs
+++ b/generators/java/generators/node/templates/buildSrc/src/main/groovy/jhipster.node-gradle-conventions.gradle.ejs
@@ -1,3 +1,6 @@
+import java.nio.file.Files
+import java.nio.file.Path
+
 plugins {
     id "com.github.node-gradle.node"
 }
@@ -9,22 +12,53 @@ if (project.hasProperty("nodeInstall")) {
         download = true
     }
 
-    // Copy local node and npm to a fixed location for npmw
-    def deleteOldNpm = tasks.register("deleteOldNpm", Delete) {
-        delete('<%- temporaryDir %>node/lib/node_modules/npm')
-    }
-    def fixedNode = tasks.register("fixedNode", Copy) {
-        from(nodeSetup)
-        into('<%- temporaryDir %>node')
-        finalizedBy(deleteOldNpm)
-    }
-    tasks.named("nodeSetup").configure { finalizedBy(fixedNode) }
+    // Link local node and npm to a fixed location for npmw.
+    // They are linked instead of copied, because a copy dereferences symlinks, and bin/npm and
+    // bin/npx must stay symlinks to lib/node_modules/npm/bin/*-cli.js: as plain extension-less
+    // files they are resolved against the wrong directory, and parsed as ESM since the closest
+    // package.json is the project one, which declares "type": "module".
+    def isWindows = System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("windows")
+    def nodeDir = tasks.named("nodeSetup").flatMap { it.nodeDir }
+    def npmDir = tasks.named("npmSetup").flatMap { it.npmDir }
+    def fixedNodeDir = layout.buildDirectory.dir("node")
 
-    def fixedNpm = tasks.register("fixedNpm", Copy) {
-        from(npmSetup)
-        into('<%- temporaryDir %>node')
+    // Link (in the fixed location) to target pairs, relative targets being resolved from the link
+    def nodeLinks = { ->
+        def fixed = fixedNodeDir.get().asFile.toPath()
+        def links = [:]
+        if (isWindows) {
+            links[fixed.resolve("bin/node.exe")] = nodeDir.get().file("node.exe").asFile.toPath()
+            links[fixed.resolve("bin/npm.cmd")] = npmDir.get().file("npm.cmd").asFile.toPath()
+            links[fixed.resolve("bin/npx.cmd")] = npmDir.get().file("npx.cmd").asFile.toPath()
+            links[fixed.resolve("lib/node_modules/npm")] = npmDir.get().dir("node_modules/npm").asFile.toPath()
+        } else {
+            links[fixed.resolve("bin/node")] = nodeDir.get().file("bin/node").asFile.toPath()
+            links[fixed.resolve("bin/npm")] = Path.of("../lib/node_modules/npm/bin/npm-cli.js")
+            links[fixed.resolve("bin/npx")] = Path.of("../lib/node_modules/npm/bin/npx-cli.js")
+            links[fixed.resolve("lib/node_modules/npm")] = npmDir.get().dir("lib/node_modules/npm").asFile.toPath()
+        }
+        links
     }
-    tasks.named("npmSetup").configure { finalizedBy(fixedNpm) }
+
+    def linkNode = tasks.register("linkNode") {
+        description = "Link the local node and npm installations to a fixed location for npmw."
+        onlyIf {
+            nodeLinks().any { link, target -> !Files.isSymbolicLink(link) || Files.readSymbolicLink(link) != target }
+        }
+        doLast {
+            nodeLinks().each { link, target ->
+                Files.createDirectories(link.parent)
+                if (Files.isDirectory(link) && !Files.isSymbolicLink(link)) {
+                    link.toFile().deleteDir()
+                } else {
+                    Files.deleteIfExists(link)
+                }
+                Files.createSymbolicLink(link, target)
+            }
+        }
+    }
+    tasks.named("nodeSetup").configure { finalizedBy(linkNode) }
+    tasks.named("npmSetup").configure { finalizedBy(linkNode) }
 }
 
 <%_ if (!skipClient) { _%>

